rear/rear
# Reorder systemd units in the rescue system and make sure syslog is started

##### Pull Request Details:

* Type: **Bug Fix**

* Impact: **Normal**

* Reference to related issue (URL):

* How was this pull request tested?
Boot of the rescue environment on RHEL 9, and logging a large number of log messages via `for i in $(seq 1 1000); do logger foo$i; done`

* Brief description of the changes in this pull request:
  * Configure network-online.target in rescue system. Services that need network to be configured can be made to be run after this target. This matches the usual systemd setup: https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/ . Intended especially for 3rd party network backup solutions where restore needs some daemon running and communicating over network in the rescue system.
  * Automated ReaR startup is now a separate script and systemd unit, not part of sysinit.service.
  * Make sysinit.service start early in rescue boot. The /etc/scripts/system-setup script is intended to start early and perform basic system initialization (according to /etc/inittab and /etc/init/rcS.conf in the pre-systemd configs). Make the sysinit.service that starts the script under systemd part of sysinit.target so that even with systemd it starts early.  Order /etc/scripts/boot, executed by rear-boot-helper.service, before sysinit.service.
  * Make multi-user target in rescue require basic target, otherwise the services and sockets required by basic target are never started (and thus logging does not work because rsyslog is not started).
  * Fix syslog: start it after rescue system init (sysinit.target) as it used to be the case without systemd: see /etc/inittab. For some reason, starting syslog in early in the rescue system boot did not work.
  * Cleanup: Remove run-system-setup.service from rescue image, it is not needed by any other systemd units, therefore it is unused - duplicates sysinit.service.

This fixes a problem with /dev/log in the rescue system: on RHEL 9 no process was reading it, as rsyslog was not being started. If during recovery a lot of log messages got generated (this happened while rerunning GRUB when there are lots of file systems, as grub-probe is very noisy), the log socket got filled up and recovery froze.

Separation of ReaR startup from sysinit.service allows it to run late in the bootup process, while sysinit.service is to be run early, and thus run network daemons after sysinit.service but before ReaR. Especially useful for network backup daemons that are needed to restore files during "rear recover".

As a side effect, I am now seeing error messages from systemd during "rear recover":
```
Starting D-Bus System Message Bus   
... 
[        FAILED   ] Failed to start D-Bus System Message Bus. 
See 'systemctl status dbus.service' for details.
```

`systemctl status dbus.service` says:
```
× dbus.service - D-Bus System Message Bus
     Loaded: loaded (/usr/lib/systemd/system/dbus.service; static)
     Active: failed (Result: exit-code) since Fri 2023-08-11 08:13:56 EDT; 2h 20min ago
TriggeredBy: × dbus.socket
    Process: 9420 ExecStartPre=/bin/dbus-uuidgen --ensure (code=exited, status=203/EXEC)
        CPU: 913us
```

This is because dbus-daemon and related utilities (dbus-uuidgen) are not present in the rescue image. dbus.service was broken even before this change, but since nothing depended on sockets.target, dbus.socket did not get activated and the problem did not appear. This problem does not seem to have adverse effect, so I am leaving it for later.
